golang/internal/dockerhelper/container.go

```diff
@@ -0,0 +1,109 @@
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/client"
+	"github.com/dyrector-io/dyrectorio/golang/internal/dogger"
+)
+
+const dockerClientTimeoutSeconds = 30
+
+func DeleteContainerByName(ctx context.Context, dog *dogger.DeploymentLogger, nameFilter string) error {
+	matchedContainer, err := GetOneContainerByName(ctx, nil, nameFilter)
+	if err != nil {
+		return fmt.Errorf("builder could not get container (%s) to remove: %s", nameFilter, err.Error())
+	}
+
+	switch matchedContainer.State {
+	case "running", "paused", "restarting":
+		if err = StopContainerByName(ctx, nil, exactMatch(nameFilter)); err != nil {
+			return fmt.Errorf("builder could not stop container (%s): %s", nameFilter, err.Error())
+		}
+		fallthrough
+	case "exited", "dead", "created":
+		if err = RemoveContainerByName(ctx, nil, exactMatch(nameFilter)); err != nil {
+			return fmt.Errorf("builder could not remove container (%s): %s", nameFilter, err.Error())
+		}
+		return nil
+	case "":
+		// when there's no container we just skip it
+		return nil
+	default:
+		return fmt.Errorf("builder could not determine the state (%s) of the container (%s) for deletion: %s",
+			matchedContainer.State,
+			nameFilter,
+			err.Error())
+	}
+}
+
+func StopContainerByName(ctx context.Context, dog *dogger.DeploymentLogger, containerName string) error {
+	cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		panic(err)
+	}
+
+	timeoutValue := (time.Duration(dockerClientTimeoutSeconds) * time.Second)
+	if err := cli.ContainerStop(ctx, containerName, &timeoutValue); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Matches one
+func RemoveContainerByName(ctx context.Context, dog *dogger.DeploymentLogger, nameFilter string) error {
+	cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		panic(err)
+	}
+
+	if err := cli.ContainerRemove(ctx, exactMatch(nameFilter), types.ContainerRemoveOptions{}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Check the existence of containers, then return it
+func GetContainersByName(ctx context.Context, nameFilter string) ([]types.Container, error) {
+	cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		panic(err)
+	}
+
+	filter := filters.NewArgs()
+	filter.Add("name", nameFilter)
+
+	containers, err := cli.ContainerList(ctx, types.ContainerListOptions{All: true, Filters: filter})
+	if err != nil {
+		return []types.Container{}, err
+	}
+
+	return containers, err
+}
+
+// Using exact match!
+func GetOneContainerByName(ctx context.Context, dog *dogger.DeploymentLogger, nameFilter string) (types.Container, error) {
+	containers, err := GetContainersByName(ctx, exactMatch(nameFilter))
+	if err != nil {
+		return types.Container{}, err
+	}
+
+	switch len(containers) {
```

Comment:
`ContainerListOptions` there is a Limit int attribute, can't we workaround this with the limit?

Reply:
Thing is when we do a "find me THAT one container" type of operation we shouldn't use limits. Issue is if the search filter isn't specific enough we could end up with more than one container, and we making space for doing an unwanted operation on a container we do not care about. I'd rather rule out the chance or opportunity to work on a wrong container.

Reply:
@polaroi8d what do you think about this issue?

Reply:
Got it, but what if the search filter isn't specific enough, what guarantees that I need the first container in the list? Because you get the first element from the list, didn't see this is enough. 

Code snippet:
```
	case 1:
		return containers[0], nil
```

Reply:
That's the idea, it's another safety to search for every matching container, and if we have more than one, we return error.
```
	switch len(containers) {
	// Ideal, where we have just one matching
	case 1:
		return containers[0], nil
	// Ideal, where we have no match
	case 0:
		return types.Container{}, nil
	// Not ideal at all, when the search pattern matched multiple, so it wasn't an exact match, this should not happen, but I'd like to keep it for safety
	default:
		return types.Container{}, fmt.Errorf("more than one matching container")
	}
```

When the search input is "random-container" and we return random-container, random-container2, random-container3 etc; this shouldn't happen because we use exact matching pattern (which isn't the default somehow...) earlier in the func.

MFW when docker does not use exact matching by default: https://static.guim.co.uk/sys-images/Guardian/Pix/pictures/2014/10/30/1414670814763/1956-MOBY-DICK-012.jpg

By all means, this code isn't that necessary but an extra safety measure.

Reply:
Ahh okay, thanks for the clarification.